libcfnet/tls_generic.c

```diff
@@ -868,6 +873,20 @@ void TLSSetDefaultOptions(SSL_CTX *ssl_ctx, const char *min_version)
         Log(LOG_LEVEL_VERBOSE,
             "Setting minimum acceptable TLS version: 1.2");
     }
+    else if (strcmp(min_version, "1.3") == 0)
+    {
+        if (strcmp(compiletime_min_version, "1.0") == 0 ||
+            strcmp(compiletime_min_version, "1.1") == 0 ||
+            strcmp(compiletime_min_version, "1.2") == 0)
+        {
+            Log(LOG_LEVEL_WARNING, "Minimum requested TLS version is %s,"
+                " however because of old OpenSSL version it is set to: %s",
+                min_version, compiletime_min_version);
```

Comment:
I don't think this should be allowed (potential downgrade attack).

If I specify TLS 1.3 to be used (e.g. for compliance reasons), I don't want CFEngine to simply pick a different version. It should stop, and let me choose whether to upgrade my openssl or to compromise and downgrade to 1.2. But this should be done explicitly.

Reply:
I agree, I was just following the same behavior that was used for older versions. Shall I change this in all the above versions too? The same logic applies there. However, using such logic only for 1.3 doesn't really make much sense to me.

Reply:
I think it should be changed for all versions. Though it is a bit interesting that it used to be like this too - can we find out some info on why? Are there commit logs or can we reach out to the developer doing it?

Reply:
@nickanderson does the above ring any bells?

Reply:
I will reach out and ask @jimis for any historical context.

Reply:
commit that introduced it: 2719ce18c2cd6ba2f192b013db11be003675cf76

If I recall, this code exists to warn if for example you are trying to turn off TLS 1.2, but the openssl library does not even have defined `SSL_OP_NO_TLSv1_2`. Apparently you can't turn off TLS 1.2 if the openssl library thinks that 1.1 is the most current TLS version.

The above commit introduces a dummy (of value 0x0) define that will apparently not turn off anything, but at least the program will compile and keep accepting connections, and the important WARNING will be printed prompting the sysadmin to recompile with recent openssl.

Reply:
> Apparently you can't turn off TLS 1.2 if the openssl library thinks that 1.1 is the most current TLS version.

That makes sense (I checked that the `NO_TLSv` macros have been being added to OpenSSL together with the support for the same particular version of TLS), but then I don't understand the logic of this:
```
#ifndef  SSL_OP_NO_TLSv1_1
#   define SSL_OP_NO_TLSv1_1 0x0                                /* nop */
    compiletime_min_version = "1.1";
```
which translates to *"if OpenSSL doesn't know about TLS 1.1, we set the minimal version to 1.1"*. And if user requests 1.2. or newer as the minimal version, we tell them 1.1 is the minimal version being used because of old OpenSSL (which doesn't support that version).

Shouldn't the user in such a case get a warning (or maybe an error) that the minimal requested version is higher than the maximal version supported by their OpenSSL library?

Reply:
It does seem like we should be erroring that up to the user, or in this case, warning since the "error" is handled.